apis/python/src/tiledbsoma/_dataframe.py

```diff
@@ -517,52 +539,100 @@ def _canonicalize_schema(
 def _build_tiledb_schema(
     schema: pa.Schema,
     index_column_names: Sequence[str],
+    domain: Optional[Sequence[Optional[Tuple[Any, Any]]]],
     tiledb_create_options: TileDBCreateOptions,
     context: SOMATileDBContext,
 ) -> tiledb.ArraySchema:
     """Converts an Arrow schema into a TileDB ArraySchema for creation."""
+
+    if domain is None:
+        domain = tuple(None for _ in index_column_names)
+    else:
+        ndom = len(domain)
+        nidx = len(index_column_names)
+        if ndom != nidx:
+            raise ValueError(
+                f"if domain is specified, it must have the same length as index_column_names; got {ndom} != {nidx}"
+            )
+
     dims = []
-    for index_column_name in index_column_names:
+    for index_column_name, slot_domain in zip(index_column_names, domain):
         pa_type = schema.field(index_column_name).type
         dtype = _arrow_types.tiledb_type_from_arrow_type(
             pa_type, is_indexed_column=True
         )
-        domain: Tuple[Any, Any]
-        if isinstance(dtype, str):
-            domain = None, None
+
+        if slot_domain is not None:
```

Comment:
It might be worth pulling this whole bit where you figure out the real domain into its own function.